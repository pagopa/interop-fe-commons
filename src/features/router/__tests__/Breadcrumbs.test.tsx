import React from 'react'
import { generateTestingRoutes, renderRoutes } from './common.mocks'
import { createMemoryHistory } from 'history'
import userEvent from '@testing-library/user-event'
import { expectTypeOf } from 'vitest'

const {
  routes,
  reactRouterDOMRoutes,
  components: { Breadcrumbs },
} = generateTestingRoutes()

const {
  reactRouterDOMRoutes: reactRouterDOMLocalizedRoutes,
  components: { Breadcrumbs: LocalizedBreadcrumbs },
} = generateTestingRoutes({ languages: ['it', 'en'] })

const routeLabels = {
  HOME: 'home',
  PAGE_1: 'page-1',
  PAGE_2: 'page-2',
  DYNAMIC_PAGE_1: 'dynamic-page-1',
  DYNAMIC_PAGE_2: 'dynamic-page-2',
  REDIRECT_1: 'redirect-1',
  REDIRECT_2: 'redirect-2',
} as const

const renderBreadcrumb = (startingRoute = '/page-1/test-id') => {
  const history = createMemoryHistory()
  history.push(startingRoute)
  const screen = renderRoutes(
    <Breadcrumbs routeLabels={routeLabels} />,
    reactRouterDOMRoutes,
    history
  )
  return { ...screen, history }
}

const renderLocalizedBreadcrumb = (startingRoute = '/it/page-1/test-id') => {
  const history = createMemoryHistory()
  history.push(startingRoute)
  const screen = renderRoutes(
    <LocalizedBreadcrumbs routeLabels={routeLabels} />,
    reactRouterDOMLocalizedRoutes,
    history
  )
  return { ...screen, history }
}

describe('Breadcrumbs', () => {
  it('should match the snapshot', () => {
    const { baseElement } = renderBreadcrumb()
    expect(baseElement).toMatchSnapshot()
  })

  it('should match the snapshot (localized)', () => {
    const history = createMemoryHistory()
    history.push('/page-1/test-id')
    const { baseElement } = renderLocalizedBreadcrumb()
    expect(baseElement).toMatchSnapshot()
  })

  it('should not render if the route have less than 2 parents', () => {
    const { container } = renderBreadcrumb('/')
    expect(container).toBeEmptyDOMElement()
  })

  it('should navigate to the correct route when clicking on a breadcrumb link', async () => {
    const { getByRole, history } = renderBreadcrumb()
    const page1Link = getByRole('link', { name: 'page-1' })
    const user = userEvent.setup()
    await user.click(page1Link)
    expect(history.location.pathname).toEqual('/page-1')
    const homeLink = getByRole('link', { name: 'home' })
    await user.click(homeLink)
    expect(history.location.pathname).toEqual('/')
  })

  it('should navigate to the correct route when clicking on a breadcrumb link (localized)', async () => {
    const { getByRole, history } = renderLocalizedBreadcrumb()
    const page1Link = getByRole('link', { name: 'page-1' })
    const user = userEvent.setup()
    await user.click(page1Link)
    expect(history.location.pathname).toEqual('/it/page-1')
    const homeLink = getByRole('link', { name: 'home' })
    await user.click(homeLink)
    expect(history.location.pathname).toEqual('/it/')
  })

  it('should have the correct routeLabels type', () => {
    type RouteLabels = Parameters<typeof Breadcrumbs>['0']['routeLabels']

    expectTypeOf<RouteLabels>().toMatchTypeOf<{ [K in keyof typeof routes]: string | false }>()
  })
})
