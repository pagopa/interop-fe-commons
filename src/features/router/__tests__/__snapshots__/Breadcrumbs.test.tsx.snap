// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumbs > should match the snapshot (localized) 1`] = `
<body>
  <div>
    <nav
      class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5jrdvz-MuiTypography-root-MuiBreadcrumbs-root"
    >
      <ol
        class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
      >
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-159dm8u-MuiTypography-root-MuiLink-root"
            href="/it/"
          >
            home
          </a>
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-159dm8u-MuiTypography-root-MuiLink-root"
            href="/it/page-1"
          >
            page-1
          </a>
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <span>
            dynamic-page-1
          </span>
        </li>
      </ol>
    </nav>
  </div>
</body>
`;

exports[`Breadcrumbs > should match the snapshot 1`] = `
<body>
  <div>
    <nav
      class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5jrdvz-MuiTypography-root-MuiBreadcrumbs-root"
    >
      <ol
        class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
      >
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-159dm8u-MuiTypography-root-MuiLink-root"
            href="/"
          >
            home
          </a>
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-159dm8u-MuiTypography-root-MuiLink-root"
            href="/page-1"
          >
            page-1
          </a>
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <span>
            dynamic-page-1
          </span>
        </li>
      </ol>
    </nav>
  </div>
</body>
`;
